xrtl/port/common/ui/timer_display_link.cc

```diff
@@ -58,33 +70,61 @@ void TimerDisplayLink::SetupTimer() {
+    // As fast as possible.
+    frame_time_micros_ = std::chrono::microseconds(0);
+  } else if (preferred_frames_per_second_ == kMaxDisplayRate) {
+    // Fixed at max display rate.
+    frame_time_micros_ = std::chrono::microseconds(static_cast<uint64_t>(
+        (1.0 / max_frames_per_second_) * 1000.0 * 1000.0));
+  } else {
+    // An actual FPS value.
+    frame_time_micros_ = std::chrono::microseconds(static_cast<uint64_t>(
+        (1.0 / preferred_frames_per_second_) * 1000.0 * 1000.0));
+  }
 
-  VLOG(1) << "DisplayLink started with interval " << frame_time_micros.count();
+  VLOG(1) << "DisplayLink started with rate " << preferred_frames_per_second_
+          << ", interval " << frame_time_micros_.count();
 
   // Schedule the timer and fire one callback immediately.
-  timer_task_ = message_loop_->DeferRepeating(
-      &pending_task_list_,
-      [this]() {
-        // We hold this lock for the duration of the callback so as to prevent
-        // the instance from being deleted from under us.
-        std::lock_guard<std::recursive_mutex> lock(mutex_);
-        if (!callback_) {
-          // Cancelled from another thread. Ignore.
-          return;
-        }
-
-        // Query frame time. Real APIs get real times, fake APIs like us get
-        // this.
-        std::chrono::microseconds timestamp_utc_micros =
-            SystemClock::default_clock()->now_utc_micros();
-
-        // Issue callback; note that it may call display link methods.
-        callback_(timestamp_utc_micros);
-      },
-      std::chrono::milliseconds(0),
-      std::chrono::duration_cast<std::chrono::milliseconds>(frame_time_micros));
+  timer_task_ = message_loop_->Defer(
+      &pending_task_list_, [this]() { Tick(); },
+      std::chrono::duration_cast<std::chrono::milliseconds>(
+          frame_time_micros_));
+}
+
+void TimerDisplayLink::Tick() {
+  // Query frame time. Real APIs get real times, fake APIs like us get
+  // this.
+  std::chrono::microseconds timestamp_utc_micros =
+      SystemClock::default_clock()->now_utc_micros();
+
+  // We hold this lock for the duration of the callback so as to prevent
+  // the instance from being deleted from under us.
+  {
+    std::lock_guard<std::recursive_mutex> lock(mutex_);
+    if (!callback_) {
+      // Cancelled from another thread. Ignore.
+      return;
+    }
+
+    // Issue callback; note that it may call display link methods.
+    callback_(timestamp_utc_micros);
+  }
+
+  // Schedule another tick. We make sure to compensate for the amount of time
+  // we've spent in the callback.
+  std::chrono::microseconds callback_duration_micros =
+      SystemClock::default_clock()->now_utc_micros() - timestamp_utc_micros;
+  std::chrono::microseconds delay_micros;
+  if (callback_duration_micros < frame_time_micros_) {
+    delay_micros = frame_time_micros_ - callback_duration_micros;
+  } else {
+    delay_micros = std::chrono::microseconds(1000);
```

Comment:
a 0hz timer would flood the message loop and make the app non-responsive. adding comment.